src/transaction/transaction_group_complete_manager.hpp

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (C) 2008 Search Solution Corporation. All rights reserved by Search Solution.
+ *
+ *   This program is free software; you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation; either version 2 of the License, or
+ *   (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+//
+// Manager of completed group
+//
+
+#ifndef _GROUP_COMPLETE_MANAGER_HPP_
+#define _GROUP_COMPLETE_MANAGER_HPP_
+
+#include "transaction_complete_manager.hpp"
+#include <atomic>
+#include <condition_variable>
+#include <mutex>
+
+namespace cubtx
+{
+  /* Group state. */
+  enum GROUP_STATE
+  {
+    GROUP_CLOSED = 0x01, /* Group closed. No other transaction can be included in a closed group. */
+    GROUP_MVCC_COMPLETED = 0x02, /* MVCC completed. */
+    GROUP_LOGGED = 0x04, /* Group log added. */
+    GROUP_COMPLETED = 0x08  /* Group completed. */
+  };
+
+  //
+  // group_complete_manager is the common interface used by complete managers based on grouping the commits
+  //
+  class group_complete_manager : public complete_manager
+  {
+    public:
+      group_complete_manager ()
+	: m_current_group_id (1)
+	, m_latest_closed_group_id (0)
+	, m_latest_closed_group_state (GROUP_CLOSED | GROUP_MVCC_COMPLETED | GROUP_LOGGED | GROUP_COMPLETED)
+      {
+
+      }
+      ~group_complete_manager () override = default;
+
+      id_type register_transaction (int tran_index, MVCCID mvccid, TRAN_STATE state) override final;
+
+      void wait_for_complete_mvcc (id_type group_id) override final;
+
+      void wait_for_complete (id_type group_id) override final;
+
+      void wait_for_logging (id_type group_id) override final;
+
+    protected:
+      bool close_current_group ();
+
+      virtual bool can_close_current_group () = 0;
+
+      virtual void prepare_complete (THREAD_ENTRY *thread_p) = 0;
+
+      virtual void do_complete (THREAD_ENTRY *thread_p) = 0;
+
+      void notify_group_mvcc_complete (const tx_group &closed_group);
+      void notify_group_logged ();
+      void notify_group_complete ();
+
+      bool is_latest_closed_group_mvcc_completed ();
+      bool is_latest_closed_group_logged ();
+      bool is_latest_closed_group_completed ();
+
+      bool is_current_group_empty ();
+
+      tx_group & get_last_closed_group ();
```

Comment:
```suggestion
      const tx_group & get_last_closed_group () const;
```

Should get_last_closed_group protect m_latest_closed_group from changes?